mlflow/transformers.py

```diff
@@ -0,0 +1,789 @@
+                  deep learning execution framework dependency requirements.
+    :return: A string representing the engine type of the model being used (one of 'tf', 'flax', or
+             'torch').
+    """
+    engine = _get_engine_type(model)
+    if engine == "torch":
+        return ["torch", "torchvision"]
+    else:
+        return [engine]
+
+
+def get_default_pip_requirements(model) -> List[str]:
+    """
+    :param model: The model instance to be saved in order to provide the required underlying
+                  deep learning execution framework dependency requirements. Note that this must
+                  be the actual model instance and not a Pipeline.
+    :return: A list of default pip requirements for MLflow Models that have been produced with the
+             ``transformers`` flavor. Calls to :py:func:`save_model()` and :py:func:`log_model()`
+             produce a pip environment that contain these requirements at a minimum.
+    """
+    from transformers import TFPreTrainedModel, FlaxPreTrainedModel, PreTrainedModel
+
+    if not isinstance(model, (TFPreTrainedModel, FlaxPreTrainedModel, PreTrainedModel)):
+        raise MlflowException(
+            "The supplied model type is unsupported. The model must be one of: PreTrainedModel, "
+            "TFPreTrainedModel, or FlaxPreTrainedModel"
+        )
+    base_reqs = ["transformers"]
+    base_reqs.extend(_model_packages(model))
+    return [_get_pinned_requirement(module) for module in base_reqs]
+
+
+def _validate_transformers_model_dict(transformers_model):
+    """
+    Validator for a submitted save dictionary for the transformers model. If any additional keys
+    are provided, raise to indicate which invalid keys were submitted.
+    """
+    if isinstance(transformers_model, dict):
+        invalid_keys = [key for key in transformers_model.keys() if key not in _SUPPORTED_SAVE_KEYS]
+        if invalid_keys:
+            raise MlflowException(
+                "Invalid dictionary submitted for 'transformers_model'. The "
+                f"key(s) {invalid_keys} are not permitted. Must be one of: "
+                f"{_SUPPORTED_SAVE_KEYS}"
+            )
+        if _MODEL_KEY not in transformers_model:
+            raise MlflowException(
+                "The 'transformers_model' dictionary must have an entry for " f"{_MODEL_KEY}"
+            )
+
+
+def get_default_conda_env(model):
+    """
+    :return: The default Conda environment for MLflow Models produced with the ``transformers``
+             flavor, based on the model instance framework type of the model to be logged.
+    """
+    return _mlflow_conda_env(additional_pip_deps=get_default_pip_requirements(model))
+
+
+@experimental
+@format_docstring(LOG_MODEL_PARAM_DOCS.format(package_name=FLAVOR_NAME))
+def save_model(
+    transformers_model,
+    path: str,
+    processor=None,
+    task: Optional[str] = None,
+    model_card=None,
+    conda_env=None,
```

Comment:
Nit: Now that we're discouraging use of conda, can we move this down after `extra_pip_requirements` and also do so for `log_model()`?

Reply:
moved in both places